operator/mina/lib/src/consensus_state.rs

```diff
@@ -1,47 +1,174 @@
-            return LongerChainResult::Candidate;
+            return ChainResult::Candidate;
         } else if hash_last_vrf(candidate) == hash_last_vrf(tip) {
             // compare consensus state hashes lexicographically
             if hash_state(candidate) > hash_state(tip) {
-                return LongerChainResult::Candidate;
+                return ChainResult::Candidate;
             }
         }
     }
 
-    LongerChainResult::Bridge
+    ChainResult::Bridge
+}
+
+/// Returns true if the fork is short-range, else the fork is long-range.
+fn is_short_range(candidate: &MinaProtocolState, tip: &MinaProtocolState) -> Result<bool, String> {
+    // TODO(xqft): verify constants are correct
+    if tip.body.constants != candidate.body.constants {
+        return Err("Protocol constants on candidate and tip state are not equal".to_string());
+    }
+    let slots_per_epoch = tip.body.constants.slots_per_epoch.as_u32();
+
+    let candidate = &candidate.body.consensus_state;
+    let tip = &tip.body.consensus_state;
+
+    let check = |s1: &MinaConsensusState, s2: &MinaConsensusState| {
+        let s2_epoch_slot = s2.global_slot() % slots_per_epoch;
+        if s1.epoch_count.as_u32() == s2.epoch_count.as_u32() + 1
+            && s2_epoch_slot >= slots_per_epoch * 2 / 3
+        {
+            s1.staking_epoch_data.lock_checkpoint == s2.next_epoch_data.lock_checkpoint
+        } else {
+            false
+        }
+    };
+
+    Ok(if candidate.epoch_count == tip.epoch_count {
+        candidate.staking_epoch_data.lock_checkpoint == tip.staking_epoch_data.lock_checkpoint
+    } else {
+        check(candidate, tip) || check(tip, candidate)
+    })
+}
+
+fn relative_min_window_density(candidate: &MinaProtocolState, tip: &MinaProtocolState) -> u32 {
+    let candidate = &candidate.body.consensus_state;
+    let tip = &tip.body.consensus_state;
+
+    let max_slot = max(candidate.global_slot(), tip.global_slot());
+
+    if max_slot < GRACE_PERIOD_END {
+        return candidate.min_window_density.as_u32();
+    }
+
+    let projected_window = {
+        let shift_count = (max_slot - candidate.global_slot() - 1).clamp(0, SUB_WINDOWS_PER_WINDOW);
+        let mut projected_window: Vec<_> = candidate
+            .sub_window_densities
+            .iter()
+            .map(|d| d.as_u32())
+            .collect();
+
+        let mut i = relative_sub_window(candidate);
+        for _ in 0..shift_count {
+            i = (i + 1) % SUB_WINDOWS_PER_WINDOW;
+            projected_window[i as usize] = 0
+        }
+
+        projected_window
+    };
+
+    let projected_window_density = projected_window.iter().sum();
+
+    min(
+        candidate.min_window_density.as_u32(),
+        projected_window_density,
+    )
 }
 
-fn hash_last_vrf(chain: &MinaStateProtocolStateValueStableV2) -> String {
+fn relative_sub_window(state: &MinaConsensusState) -> u32 {
+    (state.global_slot() / SLOTS_PER_SUB_WINDOW) % SUB_WINDOWS_PER_WINDOW
```

Comment:
We can use `state.body.constants.slots_per_sub_window` instead of `SLOTS_PER_SUB_WINDOW`.

Reply:
problem here is that the parameter is a consensus state, which doesn't have the protocol constants. Would it be better if I hardcode all the constants and check if they match with the states?

Reply:
Ok, it makes sense for now

Reply:
Leaving like this for now